flink-core/src/main/java/org/apache/flink/api/common/functions/Function.java

```diff
@@ -27,4 +28,17 @@
  * method) interfaces that can be implemented via Java 8 lambdas.
  */
 @Public
-public interface Function extends java.io.Serializable {}
+public interface Function extends java.io.Serializable {
+
+    /**
+     * Whether the function will modify the input records inplace. By default, this method return
+     * true for safety reason. If the function doesn't modify the input data, it should overwrite
+     * this method to return false for better performance.
+     *
+     * @return Whether the function will modify the input data inplace.
+     */
+    @Internal
+    default boolean willModifyInput() {
```

Comment:
It is probably not correct to say "Whether the function will modify the input records inplace". This is because this method is allowed to return true even if "the function does not modify the input records inplace".

Also note that we should be able to support object-reuse even if the function modify input, as long as the function do not cache/modify those records that have already been emitted.

```
/**
 * If this method returns true, it is guaranteed that this function will not
 * cache/modify those records that it has previously emitted.
 */
@Internal
default boolean enableObjectReuse() {
    return false;
}
```